## Data Loading and Processing

1. **Install the Required Package**:  
   First, we install the `understatapi` package, which allows you to access data from the Understat API.

2. **Import Libraries**:  
   We import `pandas` for handling and processing the data, and `understatapi` to fetch the data from Understat.

3. **Initialize the Understat Client**:  
   A client object is created to interact with the Understat API, enabling you to fetch data.

4. **Retrieve EPL Match Data**:  
   Using the client, we fetch match data for the English Premier League (EPL) for the 2024 season. This data includes match details like teams, goals, and expected goals (xG).

5. **Process the Data into a DataFrame**:  
   The raw data retrieved is organized into a pandas DataFrame. This step involves extracting specific match information, like the match ID, home/away teams, goals scored, and xG values.

6. **Expand Nested Data into Separate Columns**:  
   Some columns contain nested information (e.g., home/away team details). These nested fields are unpacked into individual columns, making it easier to work with the data.

7. **Select Key Columns**:  
   Finally, We select the most relevant columns for analysis, like match ID, match datetime, home and away team names, goals scored, and expected goals (xG).

In [ ]:
!pip install understatapi --quiet
import pandas as pd
import understatapi

client = understatapi.UnderstatClient()

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 397.2/397.2 kB 4.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 2.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.5/147.5 kB 8.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 178.7/178.7 kB 9.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.2/61.2 kB 3.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 138.1/138.1 kB 7.8 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests==2.32.3, but you have requests 2.25.1 which is incompatible.
tweepy 4.15.0 requires requests<3,

In [ ]:
# Data from EPL to retrieve a list of matches
league_data = client.league(league="EPL").get_match_data(season="2024")
league_data[0]
#df_league = pd.DataFrame(league_data)
#df_league.head()

{'id': '26602',
 'isResult': True,
 'h': {'id': '89', 'title': 'Manchester United', 'short_title': 'MUN'},
 'a': {'id': '228', 'title': 'Fulham', 'short_title': 'FLH'},
 'goals': {'h': '1', 'a': '0'},
 'xG': {'h': '2.04268', 'a': '0.418711'},
 'datetime': '2024-08-16 19:00:00',
 'forecast': {'w': '0.8069', 'd': '0.1489', 'l': '0.0442'}}

In [ ]:
import pandas as pd

# Assuming league_data is already populated as in your previous code
# league_data = client.player("Manchester United").get_match_data(season="2024")

# Create a list of dictionaries with the desired columns
extracted_data = []
for match in league_data:
    extracted_data.append({
        'ID': match['id'],
        'h': match['h'],
        'a': match['a'],
        'datetime': match['datetime'],
        'goals': match['goals'],
        'xG': match['xG']
    })

# Create a pandas DataFrame from the extracted data
df_matches = pd.DataFrame(extracted_data)

# Display the DataFrame
df_matches.head()

,ID,h,a,datetime,goals,xG
0,26602,"{'id': '89', 'title': 'Manchester United', 'sh...","{'id': '228', 'title': 'Fulham', 'short_title'...",2024-08-16 19:00:00,"{'h': '1', 'a': '0'}","{'h': '2.04268', 'a': '0.418711'}"
1,26603,"{'id': '285', 'title': 'Ipswich', 'short_title...","{'id': '87', 'title': 'Liverpool', 'short_titl...",2024-08-17 11:30:00,"{'h': '0', 'a': '2'}","{'h': '0.342601', 'a': '3.92906'}"
2,26604,"{'id': '83', 'title': 'Arsenal', 'short_title'...","{'id': '229', 'title': 'Wolverhampton Wanderer...",2024-08-17 14:00:00,"{'h': '2', 'a': '0'}","{'h': '1.6283', 'a': '0.575835'}"
3,26605,"{'id': '72', 'title': 'Everton', 'short_title'...","{'id': '220', 'title': 'Brighton', 'short_titl...",2024-08-17 14:00:00,"{'h': '0', 'a': '3'}","{'h': '0.405325', 'a': '1.79083'}"
4,26606,"{'id': '86', 'title': 'Newcastle United', 'sho...","{'id': '74', 'title': 'Southampton', 'short_ti...",2024-08-17 14:00:00,"{'h': '1', 'a': '0'}","{'h': '0.433489', 'a': '1.95483'}"


In [ ]:
# Extract data from the 'h' column
h_data = df_matches['h'].apply(pd.Series)
h_data = h_data.add_prefix('h_')  # Add prefix to avoid column name conflicts

# Extract data from the 'a' column
a_data = df_matches['a'].apply(pd.Series)
a_data = a_data.add_prefix('a_')  # Add prefix to avoid column name conflicts

# Extract data from the 'goals' column
goals_data = df_matches['goals'].apply(pd.Series)
goals_data = goals_data.add_prefix('goals_')  # Add prefix

# Extract data from the 'xg' column
xg_data = df_matches['xG'].apply(pd.Series)
xg_data = xg_data.add_prefix('xG_')  # Add prefix


# Concatenate the extracted data with the original DataFrame
df_expanded = pd.concat([df_matches, h_data, a_data], axis=1)
# Concatenate with the existing df_expanded
df_expanded = pd.concat([df_expanded, goals_data, xg_data], axis=1)

# Drop the original 'h' and 'a' columns
df_expanded = df_expanded.drop(columns=['h', 'a'])

# Drop the original 'goals' and 'xg' columns
df_expanded = df_expanded.drop(columns=['goals', 'xG'])

df_expanded.head()

,ID,datetime,h_id,h_title,h_short_title,a_id,a_title,a_short_title,goals_h,goals_a,xG_h,xG_a
0,26602,2024-08-16 19:00:00,89,Manchester United,MUN,228,Fulham,FLH,1,0,2.04268,0.418711
1,26603,2024-08-17 11:30:00,285,Ipswich,IPS,87,Liverpool,LIV,0,2,0.342601,3.92906
2,26604,2024-08-17 14:00:00,83,Arsenal,ARS,229,Wolverhampton Wanderers,WOL,2,0,1.6283,0.575835
3,26605,2024-08-17 14:00:00,72,Everton,EVE,220,Brighton,BRI,0,3,0.405325,1.79083
4,26606,2024-08-17 14:00:00,86,Newcastle United,NEW,74,Southampton,SOU,1,0,0.433489,1.95483


In [ ]:
# Select the desired columns
team_data = df_expanded[['ID', 'datetime', 'h_title', 'a_title', 'goals_h', 'goals_a', 'xG_h', 'xG_a']]
team_data.head()

,ID,datetime,h_title,a_title,goals_h,goals_a,xG_h,xG_a
0,26602,2024-08-16 19:00:00,Manchester United,Fulham,1,0,2.04268,0.418711
1,26603,2024-08-17 11:30:00,Ipswich,Liverpool,0,2,0.342601,3.92906
2,26604,2024-08-17 14:00:00,Arsenal,Wolverhampton Wanderers,2,0,1.6283,0.575835
3,26605,2024-08-17 14:00:00,Everton,Brighton,0,3,0.405325,1.79083
4,26606,2024-08-17 14:00:00,Newcastle United,Southampton,1,0,0.433489,1.95483


#### Understanding Expected Goals (xG)
##### What is xG?
* xG measures the quality of scoring chances by assigning each shot a probability of becoming a goal. Higher xG = better chances.

##### Why it matters:

* More reliable than actual goals for predicting future performance

* Shows which teams create the best chances (not just lucky scorers)

## PySpark Task
### 1. Setting Up PySpark


In [ ]:
# Install PySpark and Java
!apt-get install openjdk-8-jdk-headless -qq > /dev/null
!pip install -q pyspark

# Set up environment variables
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"

# Start a SparkSession
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("Football Analysis").getOrCreate()

### 2. Low-Level: RDD Operations


In [ ]:
# Create a PySpark DataFrame from the pandas DataFrame 'team_data'
football_df = spark.createDataFrame(team_data)

# Cast xG columns to float type
football_df = football_df.withColumn("xG_h", football_df["xG_h"].cast("float"))
football_df = football_df.withColumn("xG_a", football_df["xG_a"].cast("float"))

# Show the first few rows
football_df.show(5)

# Convert DataFrame to RDD
football_rdd = football_df.rdd

# Count total matches
total_matches = football_rdd.count()
print(f"Total matches: {total_matches}")

# Get matches with high xG difference (exciting matches)
exciting_matches = football_rdd.filter(
    lambda row: abs(row["xG_h"] - row["xG_a"]) > 2
).map(
    lambda row: (row["h_title"], row["a_title"], row["xG_h"], row["xG_a"])
)

print("\nExciting matches (xG difference > 2):")
for match in exciting_matches.take(5):
    print(f"{match[0]} vs {match[1]}: {match[2]:.2f}-{match[3]:.2f} xG")

+-----+-------------------+-----------------+--------------------+-------+-------+--------+--------+
|   ID|           datetime|          h_title|             a_title|goals_h|goals_a|    xG_h|    xG_a|
+-----+-------------------+-----------------+--------------------+-------+-------+--------+--------+
|26602|2024-08-16 19:00:00|Manchester United|              Fulham|      1|      0| 2.04268|0.418711|
|26603|2024-08-17 11:30:00|          Ipswich|           Liverpool|      0|      2|0.342601| 3.92906|
|26604|2024-08-17 14:00:00|          Arsenal|Wolverhampton Wan...|      2|      0|  1.6283|0.575835|
|26605|2024-08-17 14:00:00|          Everton|            Brighton|      0|      3|0.405325| 1.79083|
|26606|2024-08-17 14:00:00| Newcastle United|         Southampton|      1|      0|0.433489| 1.95483|
+-----+-------------------+-----------------+--------------------+-------+-------+--------+--------+
only showing top 5 rows

Total matches: 380

Exciting matches (xG difference > 2):
Ipswich 

##### Most exciting games
These games had the biggest difference in chance quality (xG difference > 2):

* Ipswich 0-2 Liverpool (xG: 0.34 vs 3.93): Liverpool dominated with 12x better chances

* Man City 3-2 Ipswich (xG: 3.08 vs 0.48): City created 6x better chances despite close score

* Southampton 0-1 Newcastle (xG: 1.95 vs 0.43): Southampton had 4x better chances but lost

### 3. Mid-Level: DataFrame and SQL Operations
#### 3.1) Team Performance Analysis

In [ ]:
from pyspark.sql.functions import col, sum, count, when, lit

# Create home and away views
home_stats = football_df.groupBy("h_title").agg(
    sum("goals_h").alias("home_goals_scored"),
    sum("goals_a").alias("home_goals_conceded"),
    sum("xG_h").alias("home_xG"),
    sum("xG_a").alias("home_xGA"),
    count(lit(1)).alias("home_matches")
)

away_stats = football_df.groupBy("a_title").agg(
    sum("goals_a").alias("away_goals_scored"),
    sum("goals_h").alias("away_goals_conceded"),
    sum("xG_a").alias("away_xG"),
    sum("xG_h").alias("away_xGA"),
    count(lit(1)).alias("away_matches")
)

# Join and calculate totals
team_stats = home_stats.join(away_stats, home_stats.h_title == away_stats.a_title) \
    .select(
        col("h_title").alias("team"),
        (col("home_goals_scored") + col("away_goals_scored")).alias("total_goals_scored"),
        (col("home_goals_conceded") + col("away_goals_conceded")).alias("total_goals_conceded"),
        (col("home_xG") + col("away_xG")).alias("total_xG"),
        (col("home_xGA") + col("away_xGA")).alias("total_xGA"),
        (col("home_matches") + col("away_matches")).alias("total_matches")
    )

# Show teams sorted by xG difference (attack strength)
team_stats.withColumn("xG_difference", col("total_xG") - col("total_xGA")) \
    .orderBy("xG_difference", ascending=False) \
    .show()

+--------------------+------------------+--------------------+------------------+------------------+-------------+--------------------+
|                team|total_goals_scored|total_goals_conceded|          total_xG|         total_xGA|total_matches|       xG_difference|
+--------------------+------------------+--------------------+------------------+------------------+-------------+--------------------+
|           Liverpool|              72.0|                30.0|  75.6054710149765|30.785567104816437|           38|  44.819903910160065|
|             Arsenal|              56.0|                26.0| 60.25794565677643| 29.68411583453417|           38|   30.57382982224226|
|         Bournemouth|              51.0|                40.0| 63.36207216978073| 45.74025642871857|           38|  17.621815741062164|
|             Chelsea|              54.0|                37.0| 61.91936016082764|46.421445809304714|           38|  15.497914351522923|
|     Manchester City|              57.0|       

##### Team Performance Insights
Top Attacking Teams (Best xG Difference):

* Liverpool (+44.8) - Creates the most high-quality chances

* Arsenal (+30.6) - Strong both in attack and defense

* Bournemouth (+17.6) - Surprisingly effective attack

Defensive Weaknesses:
* Southampton (-50.1) - Allows opponents the most high-quality chances

* Leicester (-37.7) - Struggles to prevent opponent chances

* Ipswich (-35.6) - Newly promoted team finding top level tough

#### 3.2) SQL Query for Match Outcomes


In [ ]:
# Register DataFrame as a temporary view
football_df.createOrReplaceTempView("matches")

# SQL query to find matches where the result didn't match xG expectation
surprising_results = spark.sql("""
    SELECT
        datetime,
        h_title as home_team,
        a_title as away_team,
        goals_h as home_goals,
        goals_a as away_goals,
        xG_h as home_xG,
        xG_a as away_xG,
        CASE
            WHEN goals_h > goals_a AND xG_h < xG_a THEN 'Home overperformance'
            WHEN goals_h < goals_a AND xG_h > xG_a THEN 'Away overperformance'
            ELSE 'Expected result'
        END as result_type
    FROM matches
    WHERE (goals_h > goals_a AND xG_h < xG_a) OR (goals_h < goals_a AND xG_h > xG_a)
    ORDER BY ABS(xG_h - xG_a) DESC
""")

surprising_results.show(10)

+-------------------+-----------------+-----------------+----------+----------+--------+-------+--------------------+
|           datetime|        home_team|        away_team|home_goals|away_goals| home_xG|away_xG|         result_type|
+-------------------+-----------------+-----------------+----------+----------+--------+-------+--------------------+
|2025-01-14 19:30:00|         West Ham|           Fulham|         3|         2|0.937322|2.83581|Home overperformance|
|2024-08-17 14:00:00| Newcastle United|      Southampton|         1|         0|0.433489|1.95483|Home overperformance|
|2024-10-05 14:00:00|  Manchester City|           Fulham|         3|         2| 1.36684|2.75846|Home overperformance|
|2025-04-01 20:00:00|Nottingham Forest|Manchester United|         1|         0|0.512063|1.86688|Home overperformance|
|2025-02-15 17:30:00|   Crystal Palace|          Everton|         1|         2| 2.30646|1.01565|Away overperformance|
|2024-11-02 15:00:00|      Southampton|          Everton

### 4. High-Level: Advanced Analytics
#### 4.1 Expected vs Actual Performance

In [ ]:
from pyspark.sql.functions import avg

# Calculate performance metrics
performance = football_df.withColumn("home_performance", col("goals_h") - col("xG_h")) \
    .withColumn("away_performance", col("goals_a") - col("xG_a"))

# Average performance by team (home and away)
home_perf = performance.groupBy("h_title") \
    .agg(avg("home_performance").alias("avg_home_performance"))

away_perf = performance.groupBy("a_title") \
    .agg(avg("away_performance").alias("avg_away_performance"))

# Combine results
team_performance = home_perf.join(away_perf, home_perf.h_title == away_perf.a_title) \
    .select(
        col("h_title").alias("team"),
        col("avg_home_performance"),
        col("avg_away_performance"),
        ((col("avg_home_performance") + col("avg_away_performance"))/2).alias("avg_performance")
    ) \
    .orderBy("avg_performance", ascending=False)

team_performance.show()

+--------------------+--------------------+--------------------+--------------------+
|                team|avg_home_performance|avg_away_performance|     avg_performance|
+--------------------+--------------------+--------------------+--------------------+
|Wolverhampton Wan...| 0.05621966123580933|  0.3546595573425293| 0.20543960928916932|
|   Nottingham Forest| 0.09874393939971923| 0.28471602126955986| 0.19172998033463956|
|           Tottenham| 0.19548094272613525|  -0.127841321627299| 0.03381981054941813|
|     Manchester City| 0.03844759464263916|-0.14751899614930153|-0.05453570075333118|
|           Brentford|0.061779191717505455|-0.18222873409589133|-0.06022477118919...|
|              Fulham|0.043989566154778004|-0.17244253158569336|-0.06422648271545768|
|    Newcastle United|0.008252473814146859|-0.15807311236858368| -0.0749103192772184|
|            Brighton| -0.3274266004562378| 0.14139137417078018| -0.0930176131427288|
|             Ipswich| -0.2847528699785471| 0.07881022

#### 4.2 Match Predictions Using PySpark and Understat xG Data
Here we build a basic match prediction model using your Understat data with PySpark. The expected goals (xG) metric is particularly valuable for this purpose as it provides a more stable performance indicator than actual goals.


We'll use these key techniques:

* **Feature Engineering**: Create predictive features from historical xG data

* **Model Training**: Build a classifier to predict match outcomes

* **Evaluation**: Assess model performance

##### 4.2.1 Feature Engineering

In [ ]:
from pyspark.sql import functions as F
from pyspark.sql.window import Window

# Calculate rolling averages for team performance
window_spec = Window.partitionBy("team").orderBy("date").rowsBetween(-5, -1)

# Create home and away team views
home_team_stats = football_df.select(
    F.col("h_title").alias("team"),
    F.col("datetime").alias("date"),
    F.col("xG_h").alias("xG_for"),
    F.col("xG_a").alias("xG_against"),
    F.col("goals_h").alias("goals_for"),
    F.col("goals_a").alias("goals_against"),
    F.lit("home").alias("venue")
)

away_team_stats = football_df.select(
    F.col("a_title").alias("team"),
    F.col("datetime").alias("date"),
    F.col("xG_a").alias("xG_for"),
    F.col("xG_h").alias("xG_against"),
    F.col("goals_a").alias("goals_for"),
    F.col("goals_h").alias("goals_against"),
    F.lit("away").alias("venue")
)

# Combine home and away stats
team_stats = home_team_stats.union(away_team_stats)

# Calculate rolling averages
team_features = team_stats.withColumn(
    "avg_xG_for", F.avg("xG_for").over(window_spec)
).withColumn(
    "avg_xG_against", F.avg("xG_against").over(window_spec)
).withColumn(
    "avg_goals_for", F.avg("goals_for").over(window_spec)
).withColumn(
    "avg_goals_against", F.avg("goals_against").over(window_spec)
)

# Join features back to original matches
matches_with_features = football_df.alias("matches").join(
    team_features.alias("home").select(
        "team", "date",
        "avg_xG_for", "avg_xG_against",
        "avg_goals_for", "avg_goals_against"
    ),
    (F.col("matches.h_title") == F.col("home.team")) &
    (F.col("matches.datetime") == F.col("home.date")),
    "left"
).join(
    team_features.alias("away").select(
        "team", "date",
        "avg_xG_for", "avg_xG_against",
        "avg_goals_for", "avg_goals_against"
    ),
    (F.col("matches.a_title") == F.col("away.team")) &
    (F.col("matches.datetime") == F.col("away.date")),
    "left"
)

# Define target variable (1 if home wins, 0 otherwise)
matches_with_target = matches_with_features.withColumn(
    "home_win",
    F.when(F.col("goals_h") > F.col("goals_a"), 1).otherwise(0)
)

# Show the enhanced dataset
matches_with_target.select(
    "h_title", "a_title", "goals_h", "goals_a", "home_win",
    "home.avg_xG_for", "home.avg_xG_against",
    "away.avg_xG_for", "away.avg_xG_against"
).show(5)

+-----------------+--------------------+-------+-------+--------+----------+--------------+----------+--------------+
|          h_title|             a_title|goals_h|goals_a|home_win|avg_xG_for|avg_xG_against|avg_xG_for|avg_xG_against|
+-----------------+--------------------+-------+-------+--------+----------+--------------+----------+--------------+
|          Ipswich|           Liverpool|      0|      2|       0|      NULL|          NULL|      NULL|          NULL|
| Newcastle United|         Southampton|      1|      0|       1|      NULL|          NULL|      NULL|          NULL|
|Nottingham Forest|         Bournemouth|      1|      1|       0|      NULL|          NULL|      NULL|          NULL|
|          Arsenal|Wolverhampton Wan...|      2|      0|       1|      NULL|          NULL|      NULL|          NULL|
|Manchester United|              Fulham|      1|      0|       1|      NULL|          NULL|      NULL|          NULL|
+-----------------+--------------------+-------+-------+

In [ ]:
from pyspark.sql import functions as F
from pyspark.sql.window import Window

# Calculate rolling averages with proper aliasing
window_spec = Window.partitionBy("team").orderBy("date").rowsBetween(-5, -1)

# Create home team features with clear aliases
home_features = team_features.filter(F.col("venue") == "home") \
    .select(
        F.col("team").alias("home_team"),
        F.col("date").alias("match_date"),
        F.col("avg_xG_for").alias("home_avg_xG_for"),
        F.col("avg_xG_against").alias("home_avg_xG_against"),
        F.col("avg_goals_for").alias("home_avg_goals_for"),
        F.col("avg_goals_against").alias("home_avg_goals_against")
    )

# Create away team features with clear aliases
away_features = team_features.filter(F.col("venue") == "away") \
    .select(
        F.col("team").alias("away_team"),
        F.col("date").alias("match_date"),
        F.col("avg_xG_for").alias("away_avg_xG_for"),
        F.col("avg_xG_against").alias("away_avg_xG_against"),
        F.col("avg_goals_for").alias("away_avg_goals_for"),
        F.col("avg_goals_against").alias("away_avg_goals_against")
    )

# Join with original matches using proper column references
matches_with_features = football_df.alias("matches") \
    .join(
        home_features.alias("home"),
        (F.col("matches.h_title") == F.col("home.home_team")) &
        (F.col("matches.datetime") == F.col("home.match_date")),
        "left"
    ) \
    .join(
        away_features.alias("away"),
        (F.col("matches.a_title") == F.col("away.away_team")) &
        (F.col("matches.datetime") == F.col("away.match_date")),
        "left"
    )

# Define target variable (1 if home wins, 0 otherwise)
matches_with_target = matches_with_features.withColumn(
    "home_win",
    F.when(F.col("goals_h") > F.col("goals_a"), 1).otherwise(0)
)

# Show the schema to verify column names
matches_with_target.printSchema()

root
 |-- ID: string (nullable = true)
 |-- datetime: string (nullable = true)
 |-- h_title: string (nullable = true)
 |-- a_title: string (nullable = true)
 |-- goals_h: string (nullable = true)
 |-- goals_a: string (nullable = true)
 |-- xG_h: float (nullable = true)
 |-- xG_a: float (nullable = true)
 |-- home_team: string (nullable = true)
 |-- match_date: string (nullable = true)
 |-- home_avg_xG_for: double (nullable = true)
 |-- home_avg_xG_against: double (nullable = true)
 |-- home_avg_goals_for: double (nullable = true)
 |-- home_avg_goals_against: double (nullable = true)
 |-- away_team: string (nullable = true)
 |-- match_date: string (nullable = true)
 |-- away_avg_xG_for: double (nullable = true)
 |-- away_avg_xG_against: double (nullable = true)
 |-- away_avg_goals_for: double (nullable = true)
 |-- away_avg_goals_against: double (nullable = true)
 |-- home_win: integer (nullable = false)



##### 4.2.1 Model Training
We'll use PySpark MLlib to train a classifier:

In [ ]:
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.classification import LogisticRegression
from pyspark.ml import Pipeline

from pyspark.ml.feature import VectorAssembler, Imputer


# Create feature vector, but handle null values with Imputer
# Correct feature columns - use the aliases we created
feature_cols = [
    "home_avg_xG_for", "home_avg_xG_against",
    "away_avg_xG_for", "away_avg_xG_against"
]

# Create an Imputer to replace nulls with the mean
imputer = Imputer(
    inputCols=feature_cols,
    outputCols=[f"{c}_imputed" for c in feature_cols],
    strategy="mean"  # Replace with mean, you can use other strategies too
)

# Update feature columns to use imputed values
feature_cols_imputed = [f"{c}_imputed" for c in feature_cols]

# Create feature vector using imputed columns
assembler = VectorAssembler(
    inputCols=feature_cols_imputed,
    outputCol="features"
)


# Split data
train, test = matches_with_target.randomSplit([0.8, 0.2], seed=42)

# Impute missing values in train and test data
train = imputer.fit(train).transform(train)
test = imputer.fit(test).transform(test)

# Logistic regression model
lr = LogisticRegression(
    featuresCol="features",
    labelCol="home_win",
    predictionCol="prediction",
    probabilityCol="probability"
)

# Pipeline
pipeline = Pipeline(stages=[assembler, lr])

# Train model
model = pipeline.fit(train)

# Make predictions
predictions = model.transform(test)
predictions.select("h_title", "a_title", "probability", "prediction").show(5)

+---------------+--------------------+--------------------+----------+
|        h_title|             a_title|         probability|prediction|
+---------------+--------------------+--------------------+----------+
|        Arsenal|Wolverhampton Wan...|[0.68358547815892...|       0.0|
|       West Ham|         Aston Villa|[0.68358547815892...|       0.0|
|        Chelsea|     Manchester City|[0.68358547815892...|       0.0|
|Manchester City|             Ipswich|[0.17401859647686...|       1.0|
|      Liverpool|           Brentford|[0.40457727427739...|       1.0|
+---------------+--------------------+--------------------+----------+
only showing top 5 rows

